### PR TITLE
CE-256 codepipeline for chrome-driver migration

### PR DESCRIPTION
**Pull request by gds-ahine**

### CODE REVIEW

File: concourse-chrome-driver/terraform/deployments/670214072732/module-codebuild-build-container-docker-hub.tf [hunk 1]
```diff
@@ -0,0 +1,27 @@
+module "codebuild-build-container-docker-hub-cd" {
+  source              = "github.com/alphagov/cyber-security-shared-terraform-modules//codebuild/codebuild_build_container_docker_hub"
+  stage_name          = "codepipeline-docker-image"
```
Comment (danjoneslf) [inline concourse-chrome-driver/terraform/deployments/670214072732/module-codebuild-build-container-docker-hub.tf]: The intention is that these should match the name of the codepipeline stage and actions 

```
  stage {
    name = "BuildDockerImages"

    action {
      name             = "BuildAndPushCdImage"
```

Despite the mix of cases I think it's clearer if they match.

File: concourse-chrome-driver/terraform/deployments/670214072732/variables.tf [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+variable "github_pat" {
+  description = "the github pat token to authorise access to the repo. "
+  type        = string
+  default     = "/github/pat"
+}
+
+variable "repo_name" {
+  description = "The repository on Github"
+  type        = string
+  default     = "cyber-security-concourse-base-image"
```
Comment (danjoneslf) [inline concourse-chrome-driver/terraform/deployments/670214072732/variables.tf:91]: If the repo names sent to codebuild are hard-coded - is this used?

File: concourse-chrome-driver/terraform/deployments/670214072732/action_triggers.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
+[
+  {
+      "action": "BuildAndPushCdImage", 
+      "trigger_paths": [
+        "Dockerfile",
+	"bin"
+      ]
+  }
```
Comment (danjoneslf) [inline concourse-chrome-driver/terraform/deployments/670214072732/action_triggers.json:15]: Should this have an entry for the "BuildAndPushConcourseImage" action?

File: concourse-chrome-driver/terraform/deployments/670214072732/codepipeline.tf [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      input_artifacts  = ["git_base_image", "actions_required"]
+      output_artifacts = []
+
+      configuration = {
+        PrimarySource        = "git_base_image"
+        ProjectName          = module.codebuild-build-container-docker-hub-concourse.project_name
+        EnvironmentVariables = jsonencode([{ "name" : "CHECK_TRIGGER", "value" : 1 }, { "name" : "ACTION_NAME", "value" : "BuildAndPushCdImage" }])
+      }
+    }
+  }
```
Comment (danjoneslf) [inline concourse-chrome-driver/terraform/deployments/670214072732/codepipeline.tf:118]: Need to add a stage to apply the pipeline terraform

File: concourse-chrome-driver/terraform/deployments/670214072732/module-codebuild-build-container-docker-hub.tf (hunk 1 above)
Comment (gds-ahine) [reply concourse-chrome-driver/terraform/deployments/670214072732/module-codebuild-build-container-docker-hub.tf]: Cheers. Changed to CamelCase 👍

File: concourse-chrome-driver/terraform/deployments/670214072732/module-codebuild-build-container-docker-hub.tf [hunk 5]
```diff
@@ -0,0 +1,27 @@
+module "codebuild-build-container-docker-hub-cd" {
+  source              = "github.com/alphagov/cyber-security-shared-terraform-modules//codebuild/codebuild_build_container_docker_hub"
+  stage_name          = "CodepipelineDockerImage"
```
Comment (danjoneslf) [inline concourse-chrome-driver/terraform/deployments/670214072732/module-codebuild-build-container-docker-hub.tf]: The stage is called "BuildDockerImages"

```
  stage {
    name = "BuildDockerImages"
```

File: concourse-chrome-driver/terraform/deployments/670214072732/variables.tf (hunk 2 above)
Comment (gds-ahine) [reply concourse-chrome-driver/terraform/deployments/670214072732/variables.tf:91]: The `codebuild_get_changed_file_list` we're using, requires it -:

https://github.com/alphagov/cyber-security-shared-terraform-modules/blob/master/codebuild/codebuild_get_changed_file_list/codebuild_get_changed_file_list.tf#L64

https://github.com/alphagov/cyber-security-shared-terraform-modules/blob/master/codebuild/codebuild_get_changed_file_list/variables.tf#L49

File: concourse-chrome-driver/terraform/deployments/670214072732/action_triggers.json (hunk 3 above)
Comment (gds-ahine) [reply concourse-chrome-driver/terraform/deployments/670214072732/action_triggers.json:15]: Not entirely sure but can add if you think it's required 👍

File: concourse-chrome-driver/terraform/deployments/670214072732/codepipeline.tf (hunk 4 above)
Comment (gds-ahine) [reply concourse-chrome-driver/terraform/deployments/670214072732/codepipeline.tf:118]: do we have an example of where this is used ? Thanks.

**Comment by gds-ahine** [comment]
Pipeline is now working end-to-end.

File: concourse-chrome-driver/terraform/deployments/670214072732/variables.tf [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  default     = ""
+}
+
+variable "docker_hub_credentials" {
+  description = "Name of the secret in SSM that stores the Docker Hub credentials"
+  type        = string
+  default     = "docker_hub_credentials"
+}
+
+variable "stage_name" {
```
Comment (danjoneslf) [inline concourse-chrome-driver/terraform/deployments/670214072732/variables.tf]: I don't think these should be required. These refer to the stage name and action name in the codepipeline to be passed to the individual codebuild actions.

File: concourse-chrome-driver/terraform/deployments/670214072732/module-codebuild-build-container-docker-hub.tf [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  docker_hub_password = local.docker_hub_password
+  pipeline_name       = var.pipeline_name
+  environment         = var.environment
+}
+
+module "codebuild-build-container-docker-hub-concourse" {
+  source              = "github.com/alphagov/cyber-security-shared-terraform-modules//codebuild/codebuild_build_container_docker_hub"
+  service_role_name   = var.codebuild_service_role_name
+  docker_hub_repo     = "gdscyber/concourse-chrome-driver"
+  stage_name          = "ConcourseDockerImage"
```
Comment (danjoneslf) [inline concourse-chrome-driver/terraform/deployments/670214072732/module-codebuild-build-container-docker-hub.tf]: This stage should be BuildDockerImages - It's the stage name from the codepipeline

File: concourse-chrome-driver/terraform/deployments/670214072732/variables.tf (hunk 6 above)
Comment (gds-ahine) [reply concourse-chrome-driver/terraform/deployments/670214072732/variables.tf]: Ta removed.

File: concourse-chrome-driver/terraform/deployments/670214072732/module-codebuild-build-container-docker-hub.tf (hunk 7 above)
Comment (gds-ahine) [reply concourse-chrome-driver/terraform/deployments/670214072732/module-codebuild-build-container-docker-hub.tf]: Updated

Comment (danjoneslf) [reply concourse-chrome-driver/terraform/deployments/670214072732/module-codebuild-build-container-docker-hub.tf]: It's still showing as "ConcourseDockerImage" here - have you pushed the change?

Comment (gds-ahine) [reply concourse-chrome-driver/terraform/deployments/670214072732/module-codebuild-build-container-docker-hub.tf]: Applied the TF, just testing it before I push the commits.

File: concourse-chrome-driver/terraform/deployments/670214072732/module-codebuild-get-changed-file-list.tf [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+  deployment_role_name        = "CodePipelineDeployerRole_${data.aws_caller_identity.current.account_id}"
+  codebuild_image             = var.codebuild_image
+  pipeline_name               = var.pipeline_name
+  stage_name                  = "Changes"
+  action_name                 = "GetChangedFiles"
+  environment                 = var.environment
+  github_pat                  = var.github_pat
+  repo_name                   = var.repo_name
+  docker_hub_credentials      = var.docker_hub_credentials
+  output_artifact_path        = var.output_artifact_path
```
Comment (danjoneslf) [inline concourse-chrome-driver/terraform/deployments/670214072732/module-codebuild-get-changed-file-list.tf]: This is the problem. What that module was doing before was writing the files into a custom path in the artifact bucket. What it does now is write to a named file in the designated output artifact. 

So `output_artifact_path` for `get_changed_files` should be set to `changed_files.json` or omitted. 
and `output_artifact_path` for `get_actions_required` should be set to `actions_required.json` or omitted.

The defaults should just work if you don't set the vars.

File: concourse-chrome-driver/terraform/deployments/670214072732/codepipeline.tf [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+      name             = "Source"
+      category         = "Source"
+      owner            = "AWS"
+      provider         = "CodeStarSourceConnection"
+      version          = "1"
+      output_artifacts = ["git_base_image"]
+      configuration = {
+        ConnectionArn    = "arn:aws:codestar-connections:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:connection/${var.codestar_connection_id}"
+        FullRepositoryId = "alphagov/cyber-security-concourse-base-image"
+        BranchName       = "CE-256-migrate-pipeline"
```
Comment (danjoneslf) [inline concourse-chrome-driver/terraform/deployments/670214072732/codepipeline.tf]: Don't forget to change the branch name back

**Review by danjoneslf — DISMISSED** [review]
Don't forget to change the branch name on the source stage.

File: concourse-chrome-driver/terraform/deployments/670214072732/codepipeline.tf (hunk 9 above)
Comment (gds-ahine) [reply concourse-chrome-driver/terraform/deployments/670214072732/codepipeline.tf]: done thanks 👍

**Review by alice-carr — APPROVED** [review]

**Pull request merged**